reflex/components/core/match.py

```diff
@@ -0,0 +1,243 @@
+                )
+                if not isinstance(el, (BaseVar, BaseComponent)):
+                    raise ValueError("Case element must be a var or component")
+                case_list.append(el)
+
+            match_cases.append(case_list)
+
+        return match_cases
+
+    @classmethod
+    def _validate_return_types(cls, match_cases: List[List[BaseVar]]) -> None:
+        """Validate that match cases have the same return types.
+
+        Args:
+            match_cases: The match cases.
+
+        Raises:
+            TypeError: If the return types of cases are different.
+        """
+        first_case_return = match_cases[0][-1]
+        return_type = type(first_case_return)
+
+        if types._isinstance(first_case_return, BaseComponent):
+            return_type = BaseComponent
+        elif types._isinstance(first_case_return, BaseVar):
+            return_type = BaseVar
+
+        for case in match_cases:
+            if not types._issubclass(type(case[-1]), return_type):
+                raise TypeError("match cases should have the same return types")
+
+    @classmethod
+    def _create_match_cond_var_or_component(
+        cls,
+        match_cond_var: Var,
+        match_cases: List[List[BaseVar]],
+        default: Optional[Union[BaseVar, BaseComponent]],
+    ) -> Union[Component, BaseVar]:
+        """Create and return the match condition var or component.
+
+        Args:
+            match_cond_var: The match condition.
+            match_cases: The list of match cases.
+            default: The default case.
+
+        Returns:
+            The match component wrapped in a fragment or the match var.
+
+        Raises:
+            ValueError: If the return types are not vars when creating a match var for Var types.
+        """
+        if default is None and types._issubclass(
+            type(match_cases[0][-1]), BaseComponent
+        ):
+            default = Fragment.create()
+
+        if types._issubclass(type(match_cases[0][-1]), BaseComponent):
+            return Fragment.create(
+                cls(
+                    cond=match_cond_var,
+                    match_cases=match_cases,
+                    default=default,
+                )
+            )
+
+        # Validate the match cases (as well as the default case) to have Var return types.
+        if any(
+            case for case in match_cases if not types._isinstance(case[-1], BaseVar)
+        ) or not types._isinstance(default, BaseVar):
+            raise ValueError("Return types of match cases should be Vars.")
+
+        return match_cond_var._replace(
+            _var_name=format.format_match(
+                cond=match_cond_var._var_full_name,
+                match_cases=match_cases,  # type: ignore
+                default=default,  # type: ignore
+            ),
+            _var_type=default._var_type,  # type: ignore
+            _var_is_local=False,
+            _var_full_name_needs_state_prefix=False,
+        )
```

Comment:
we also need to `merge_var_data=VarData.merge(v._var_data for v in match_cases)` to sweep up any hooks/imports from stateful vars that are part of the match set.

Reply:
Got it. Whats the best way to test this out?